File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        def diff(a, b):
            x = 0
            for i, j in zip(a, b):
                if i != j:
                    x += 1
            return x
        vis = set()
        steps = 0
        bank = set(bank)
        q = []
        q.append(start)
        while q:
            steps += 1
            size = len(q)
            temp = set()
            while size > 0:
                curr = q.pop(0)
                size -= 1
                for i in bank:
                    if diff(curr, i) == 1:
                        if i not in vis:
                            vis.add(i)
                            temp.add(i)
            if end in temp:
                return steps
            for i in temp:
                q.append(i)
        return -1

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        Set<String> ban = new HashSet<>(Arrays.asList(bank));
        Queue<String> q = new LinkedList<>();
        q.offer(start);
        Set<String> vis = new HashSet<>();
        int steps = 0;
        while(!q.isEmpty()){
            int size = q.size();
            steps++;
            while(size-- > 0){
                String curr = q.poll();
                Set<String> temp = new HashSet<>();
                for(String s : ban)
                    if(diff(curr, s) == 1 && vis.add(s))
                        temp.add(s);
                if(temp.contains(end))
                    return steps;
                for(String s : temp)
                    q.add(s);
            }
        }
        return -1;
    }
    public int diff(String a, String b){
        int d = 0;
        for(int i = 0; i < a.length(); i++)
            if(a.charAt(i) != b.charAt(i))
                d++;
        return d;
    }
}